### PR TITLE
Add a functor which can describe a face-centered field

### LINKED ISSUE #22970: Create FaceCenteredMapFunctor for face-centered fields
## Reason
In the segregated solvers of the NavierStokes module, we communicate between the pressure and velocity equations using the face fluxes/velocities. For this, we need a functor that can store values at face centers and reconstruct cell values based on those. 

## Design
Adding a FaceCenteredMapFunctor which is similar to the CellCenteredMapFunctor but overrides the evaluation methods differently.

## Impact
Enables Finite Volume fields which are face based. Makes the implementation of segregated solvers easier.

### PR DESCRIPTION
**Pull request by grmnptr**
Closes #22970

### CODE REVIEW

**Review by GiudGiud — COMMENTED** [review]
was just taking a look to see if it relied on stuff Alex is deleting in the other PR.
The answer is not really, you ll just have to delete a few lines

File: framework/src/mesh/MooseMesh.C
Comment (GiudGiud) [inline framework/src/mesh/MooseMesh.C:3538]: reformat

**Comment by grmnptr** [comment]
Thanks @GiudGiud  for taking a look in the middle of your vacation!

**Comment by grmnptr** [comment]
@GiudGiud , @lindsayad Can I get a review on this? Or should we wait until the porosity treatment is merged?

**Comment by GiudGiud** [comment]
do you need this for a study?

**Comment by grmnptr** [comment]
We ned this for the segregated solver, thought I could separate it into a different PR

**Comment by lindsayad** [comment]
Would be nice to get more coverage

**Comment by grmnptr** [comment]
Added unit tests to increase coverage!

**Comment by loganharbour** [comment]
I would like to add in here that we shouldn't bother with adding coverage to unimplemented methods. I think @grmnptr was focused on the big number, when in reality adding coverage to `mooseError("unimplemented foo")` is meaningless and a waste of time

File: framework/src/mesh/MooseMesh.C
Comment (lindsayad) [inline framework/src/mesh/MooseMesh.C:3416]:
```suggestion
  return libmesh_map_find(_elem_to_elem_info, id);
```

File: modules/navier_stokes/include/utils/FaceCenteredMapFunctor.h
Comment (lindsayad) [inline modules/navier_stokes/include/utils/FaceCenteredMapFunctor.h:105]:
```suggestion
```

File: modules/navier_stokes/src/utils/FaceCenteredMapFunctor.C
Comment (lindsayad) [inline modules/navier_stokes/src/utils/FaceCenteredMapFunctor.C:37]: Please describe a bit what `PrimitiveType` is

File: modules/navier_stokes/src/utils/FaceCenteredMapFunctor.C
Comment (lindsayad) [inline modules/navier_stokes/src/utils/FaceCenteredMapFunctor.C:47]:
```suggestion
    const Elem * const elem = elem_arg.elem;
    mooseAssert(elem, "Elem must be non-null");
```

File: modules/navier_stokes/src/utils/FaceCenteredMapFunctor.C
Comment (lindsayad) [inline modules/navier_stokes/src/utils/FaceCenteredMapFunctor.C:51]:
```suggestion
      const Elem * const neighbor = elem->neighbor_ptr(side);
```

File: modules/navier_stokes/src/utils/FaceCenteredMapFunctor.C
Comment (lindsayad) [inline modules/navier_stokes/src/utils/FaceCenteredMapFunctor.C:61]:
```suggestion
      const ValueType face_value = this->evaluate(fi);
```

File: modules/navier_stokes/src/utils/FaceCenteredMapFunctor.C
Comment (lindsayad) [inline modules/navier_stokes/src/utils/FaceCenteredMapFunctor.C:66]:
```suggestion
      for (const auto i : make_range(dim))
```

File: modules/navier_stokes/src/utils/FaceCenteredMapFunctor.C
Comment (lindsayad) [inline modules/navier_stokes/src/utils/FaceCenteredMapFunctor.C:69]:
```suggestion
        for (const auto j : make_range(dim))
```

File: modules/navier_stokes/src/utils/FaceCenteredMapFunctor.C
Comment (lindsayad) [inline modules/navier_stokes/src/utils/FaceCenteredMapFunctor.C:75]: Nice comment!

File: modules/navier_stokes/src/utils/FaceCenteredMapFunctor.C
Comment (lindsayad) [inline modules/navier_stokes/src/utils/FaceCenteredMapFunctor.C:80]:
```suggestion
    for (const auto i : make_range(dim))
```

File: modules/navier_stokes/src/utils/FaceCenteredMapFunctor.C
Comment (lindsayad) [inline modules/navier_stokes/src/utils/FaceCenteredMapFunctor.C:32]: Again very much appreciate the comment...you're doing a good job of answering my questions as I come up with them!

File: modules/navier_stokes/test/src/postprocessors/TestFaceToCellReconstruction.C
Comment (lindsayad) [inline modules/navier_stokes/test/src/postprocessors/TestFaceToCellReconstruction.C:47]:
```suggestion
  for (const Elem * elem : _fe_problem.mesh().getMesh().active_local_element_ptr_range())
```

File: modules/navier_stokes/test/src/postprocessors/TestFaceToCellReconstruction.C
Comment (lindsayad) [inline modules/navier_stokes/test/src/postprocessors/TestFaceToCellReconstruction.C:51]: This won't be the correct volume in non-XYZ  coordinate systems but for a test object this is fine

File: modules/navier_stokes/test/src/postprocessors/TestFaceToCellReconstruction.C
Comment (lindsayad) [inline modules/navier_stokes/test/src/postprocessors/TestFaceToCellReconstruction.C:35]: So I was looking through this PR wondering at the motivation for the change in what `faceInfo()` holds and the addition of the variational iterators, and I'm thinking that this is it? To avoid doing a communication after filling the truly local face elements? This is fine in this use case, but I'm concerned that users of this functor may fill a non-local map entry on one process which is local to another process which fills in a different value. This could be addressed with some parallel consistency checks, but maybe I'm being overly paranoid.

 The other consideration is: what is more expensive? Map entry computation or communication? Probably in most cases it will be the latter which would suggest your design is a good one ... but then you have to make a trade with the cost of filtering through non-local face informations when computing your owned-ranges.

Anyway this is a long monologue ... I think what you've done is probably perfectly fine. If something related to this shows up in profiling we can revert the framework level changes and do communication when building the face map functors

**Review by lindsayad — CHANGES_REQUESTED** [review]

File: modules/navier_stokes/test/src/postprocessors/TestFaceToCellReconstruction.C
Comment (lindsayad) [reply modules/navier_stokes/test/src/postprocessors/TestFaceToCellReconstruction.C:51]: Actually might as well:
```suggestion
    const Real elem_volume = _assembly.elementVolume(elem);
```
And now I'm realizing that `ElemInfo::volume()` returns wrong values for non-XYZ coordinate systems. This is undoubtedly a hold-over from Robert's work in FaceInfo ... this definitely troubles me. We should make an issue and probably remove this API because it's very dangerous

**Review by lindsayad — CHANGES_REQUESTED** [review]

**Comment by grmnptr** [comment]
Thanks for the review @lindsayad ! I'll add the modifications and think about the longer comments over the weekend!

File: modules/navier_stokes/test/src/postprocessors/TestFaceToCellReconstruction.C
Comment (lindsayad) [reply modules/navier_stokes/test/src/postprocessors/TestFaceToCellReconstruction.C:51]: I've opened https://github.com/idaholab/moose/issues/23094

**Review by lindsayad — APPROVED** [review]

**Pull request merged**
